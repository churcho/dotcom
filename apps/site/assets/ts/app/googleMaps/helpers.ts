// @ts-ignore
import stopMarker from "../../../static/images/icon-map-stop-marker.svg";
// @ts-ignore
import stationMarker from "../../../static/images/icon-map-station-marker.svg";
// @ts-ignore
import selectedStopMarker from "../../../static/images/icon-map-stop-marker-hover.svg";
// @ts-ignore
import selectedStationMarker from "../../../static/images/icon-map-station-marker-hover.svg";
// @ts-ignore
import currentLocationMarker from "../../../static/images/icon-current-location-marker.svg";
import { iconSize } from "../../../js/google-map/helpers";
import mapStyles from "../../../js/google-map/styles";
import { MarkerData, MapData } from "./__googleMaps";

export const stopIcon = (selected: boolean): string =>
  selected ? selectedStopMarker : stopMarker;
export const stationIcon = (selected: boolean): string =>
  selected ? selectedStationMarker : stationMarker;

export const buildIconFromSVG = (
  svg: string,
  size: string
): google.maps.Icon => {
  const sizeInt = iconSize(size);

  const encoded = window.btoa(svg);

  return {
    url: `data:image/svg+xml;base64, ${encoded}`,
    scaledSize: new window.google.maps.Size(sizeInt, sizeInt),
    origin: new window.google.maps.Point(0, 0),
    labelOrigin: new window.google.maps.Point(sizeInt / 2, sizeInt / 2)
  };
};

export const buildMarkerIcon = (
  data: MarkerData,
  selected: boolean
): google.maps.Icon | undefined => {
  if (data.icon) {
    if (data.icon!.includes("station")) {
      return buildIconFromSVG(stationIcon(selected), data.size);
    }
    if (data.icon!.includes("stop")) {
      return buildIconFromSVG(stopIcon(selected), data.size);
    }
    if (data.icon!.includes("current-location")) {
      return buildIconFromSVG(currentLocationMarker, data.size);
    }
  }
  return undefined;
};

export const setMapDefaults = (
  map: google.maps.Map,
  initialData: MapData
): void => {
  const options = {
    center: new window.google.maps.LatLng(
      initialData.default_center.latitude,
      initialData.default_center.longitude
    ),
    ...initialData.dynamic_options
  };
  const styles = mapStyles as google.maps.MapTypeStyle[];
  // Hide stop icons on the transit layer
  styles.push({
    featureType: "transit",
    elementType: "labels.icon",
    stylers: [{ visibility: "off" }]
  });
  map.setOptions({ styles, ...options });
  map.setZoom(initialData.zoom || 17);

  if (initialData.layers && initialData.layers.transit) {
    const layer = new window.google.maps.TransitLayer();
    layer!.setMap(map);
  }
};
