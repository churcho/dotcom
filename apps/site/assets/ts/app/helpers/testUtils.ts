// eslint-disable-next-line import/prefer-default-export
export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};
