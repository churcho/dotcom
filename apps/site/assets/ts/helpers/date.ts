// Returns a date in the form of "August 7, 2019"
export const formattedDate = (unformatted: string): string => {
  const [year, month, day] = unformatted
    .split(/-/)
    .map(part => Number.parseInt(part, 10));

  // Remember that months in JS are 0-indexed for some reason, hence "month - 1".
  const parsedDate = new Date(year, month - 1, day);
  return parsedDate.toLocaleDateString("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric",
    timeZone: "UTC"
  });
};

// Returns a time in the form of "05:16 AM"
export const shortTime = (date: string): string =>
  new Date(date).toLocaleTimeString("en-US", {
    hour: "2-digit",
    minute: "2-digit"
  });

export default formattedDate;
