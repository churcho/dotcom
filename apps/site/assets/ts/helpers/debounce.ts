// Debounce is meant to be used with event listeners. Use it to delay a call
// until after a burst of events, such as keyboard input.
//
//   usage:
//
//     const callback = (ev: Event) => {}
//     document.addEventListener("keyup", debounce(callback, 250))
//

const debounce = (callback: Function, delay: number): EventListener => {
  let timer: number | undefined;

  return (...args) => {
    window.clearTimeout(timer);
    timer = window.setTimeout(() => callback(...args), delay);
  };
};

export default debounce;
