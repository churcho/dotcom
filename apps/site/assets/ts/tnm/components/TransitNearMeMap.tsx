import React, { ReactElement } from "react";
// @ts-ignore: Not typed
import GoogleMap from "../../../js/google-map-class";

interface Props {
  initialData: any;
  mapElementId: string;
}

class TransitNearMeMap extends React.Component<Props> {
  public componentDidMount(): void {
    const { mapElementId, initialData } = this.props;
    this.map = new GoogleMap(mapElementId, initialData);
    this.map.bind(this);
    if (initialData.markers.length > 0) {
      this.tightenBounds();
    }
  }

  /* eslint-disable typescript/no-explicit-any */
  public data: any;

  public map: any;
  /* eslint-enable typescript/no-explicit-any */

  public tightenBounds(): void {
    this.map.resetBounds(["current-location", "radius-west", "radius-east"]);
  }

  // eslint-disable-next-line
  public render(): ReactElement<any> {
    return <div />;
  }
}

export default TransitNearMeMap;
