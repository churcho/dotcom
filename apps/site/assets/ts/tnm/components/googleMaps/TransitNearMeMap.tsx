import React, { ReactElement } from "react";
import Marker from "./Marker";
import { MapData, MarkerData } from "../__tnm";
import { clickMarkerAction, SelectedStopType } from "../../state";
import mapStyles from "../../../../js/google-map/styles";

interface Props {
  initialData: MapData;
  mapElementId: string;
  dispatch: Function;
  selectedStopId: SelectedStopType;
  shouldCenterMapOnSelectedStop: boolean;
}

class TransitNearMeMap extends React.Component<Props> {
  public map: google.maps.Map | null = null;

  public componentDidMount(): void {
    const { mapElementId, initialData, dispatch } = this.props;
    const mapElement = document.getElementById(mapElementId);
    if (!mapElement) return;

    const options = {
      center: new window.google.maps.LatLng(
        initialData.default_center.latitude,
        initialData.default_center.longitude
      )
    };
    const map = new window.google.maps.Map(mapElement, options);
    map.addListener("click", () => {
      dispatch(clickMarkerAction(null));
    });
    const styles = mapStyles as google.maps.MapTypeStyle[];
    // Hide stop icons on the transit layer
    styles.push({
      featureType: "transit",
      elementType: "labels.icon",
      stylers: [{ visibility: "off" }]
    });
    map.setOptions({ styles });
    map.setZoom(initialData.zoom || 17);
    this.map = map;

    this.setBounds();
    if (initialData.layers.transit) {
      const layer = new window.google.maps.TransitLayer();
      layer!.setMap(map);
    }

    this.forceUpdate();
  }

  public setBounds(): void {
    const halfMile = 0.008333;
    const { initialData } = this.props;
    const currentLocation = initialData.markers.find(
      (marker: MarkerData) => marker.id === "current-location"
    );

    if (currentLocation) {
      const east = new google.maps.LatLng(
        currentLocation!.latitude + halfMile,
        currentLocation!.longitude
      );
      const west = new google.maps.LatLng(
        currentLocation!.latitude - halfMile,
        currentLocation!.longitude
      );
      const bounds = new window.google.maps.LatLngBounds();

      bounds.extend(east);
      bounds.extend(west);
      this.map!.fitBounds(bounds);
    }
  }

  public centerMap(): void {
    const { initialData, selectedStopId } = this.props;
    const marker = initialData.markers.find(m => m.id === selectedStopId);
    this.map!.setCenter(
      new google.maps.LatLng(marker!.latitude, marker!.longitude)
    );
  }

  public render(): ReactElement<HTMLElement> | null {
    const {
      dispatch,
      selectedStopId,
      shouldCenterMapOnSelectedStop,
      initialData
    } = this.props;
    if (this.map) {
      const markers = initialData.markers.map((marker: MarkerData) => {
        const isSelected = selectedStopId === marker.id;

        return (
          <Marker
            key={marker.id}
            data={marker}
            map={this.map!}
            dispatch={dispatch}
            isSelected={isSelected}
          />
        );
      });

      // eslint-disable-next-line no-unused-expressions
      shouldCenterMapOnSelectedStop && selectedStopId && this.centerMap();

      return <>{markers}</>;
    }
    return null;
  }
}

export default TransitNearMeMap;
