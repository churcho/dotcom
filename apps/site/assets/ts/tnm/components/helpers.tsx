import React, { KeyboardEvent } from "react";

export const handleEnterKeyPress = (
  e: KeyboardEvent,
  onClick: Function
): void => {
  if (e.key === "Enter") {
    onClick();
  }
};

export const renderSvg = (className: string, svgText: string): JSX.Element => (
  <span
    className={className}
    // eslint-disable-next-line react/no-danger
    dangerouslySetInnerHTML={{ __html: svgText }}
  />
);
