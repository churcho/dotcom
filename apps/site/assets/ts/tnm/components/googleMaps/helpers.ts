// @ts-ignore
import stopMarker from "../../../../static/images/icon-map-stop-marker.svg";
// @ts-ignore
import stationMarker from "../../../../static/images/icon-map-station-marker.svg";
// @ts-ignore
import selectedStopMarker from "../../../../static/images/icon-map-stop-marker-hover.svg";
// @ts-ignore
import selectedStationMarker from "../../../../static/images/icon-map-station-marker-hover.svg";
// @ts-ignore
import currentLocationMarker from "../../../../static/images/icon-current-location-marker.svg";
import { iconSize } from "../../../../js/google-map/helpers";

import { MarkerData } from "../__tnm";

interface GoogleMapMarkerSVG {
  url: string;
  scaledSize: google.maps.Size;
  origin: google.maps.Point;
  anchor: google.maps.Point;
  labelOrigin: google.maps.Point;
}

export const stopIcon = (selected: boolean): string =>
  selected ? selectedStopMarker : stopMarker;
export const stationIcon = (selected: boolean): string =>
  selected ? selectedStationMarker : stationMarker;

export const buildIconFromSVG = (
  svg: string,
  size: string
): GoogleMapMarkerSVG => {
  const sizeInt = iconSize(size);

  const encoded = window.btoa(svg);

  return {
    url: `data:image/svg+xml;base64, ${encoded}`,
    scaledSize: new window.google.maps.Size(sizeInt, sizeInt),
    origin: new window.google.maps.Point(0, 0),
    anchor: new window.google.maps.Point(sizeInt / 2, sizeInt / 2),
    labelOrigin: new window.google.maps.Point(sizeInt / 2, sizeInt / 2)
  };
};

export const buildMarkerIcon = (
  data: MarkerData,
  selected: boolean
): GoogleMapMarkerSVG | undefined => {
  if (data.icon) {
    if (data.icon!.includes("station")) {
      return buildIconFromSVG(stationIcon(selected), data.size);
    }
    if (data.icon!.includes("stop")) {
      return buildIconFromSVG(stopIcon(selected), data.size);
    }
    if (data.icon!.includes("current-location")) {
      return buildIconFromSVG(currentLocationMarker, data.size);
    }
  }
  return undefined;
};
