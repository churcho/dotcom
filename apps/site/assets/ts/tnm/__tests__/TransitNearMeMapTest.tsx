import React from "react";
import renderer from "react-test-renderer";
import TransitNearMeMap from "../components/TransitNearMeMap";
import { createReactRoot } from "../../app/helpers/testUtils";

it("it renders without initial markers", () => {
  createReactRoot();
  const tree = renderer
    .create(
      <TransitNearMeMap
        mapElementId="test"
        dispatch={() => {}}
        selectedStopId={null}
        stopData={[]}
        selectedModes={[]}
        shouldFilterMarkers={false}
        shouldCenterMapOnSelectedStop={false}
        /* eslint-disable typescript/camelcase */
        initialData={{
          default_center: {
            latitude: 0,
            longitude: 0
          },
          zoom: 14,
          width: 630,
          height: 630,
          scale: 1,
          reset_bounds_on_update: false,
          paths: [],
          markers: [],
          dynamic_options: {},
          layers: { transit: false },
          auto_init: true,
          bound_padding: null
        }}
        /* eslint-enable typescript/camelcase */
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("it renders with initial markers", () => {
  createReactRoot();
  const tree = renderer
    .create(
      <TransitNearMeMap
        mapElementId="test"
        dispatch={() => {}}
        selectedStopId={null}
        stopData={[]}
        selectedModes={[]}
        shouldFilterMarkers={false}
        shouldCenterMapOnSelectedStop={false}
        initialData={{
          zoom: 14,
          width: 630,
          scale: 1,
          reset_bounds_on_update: false, // eslint-disable-line
          paths: [],
          /* eslint-disable typescript/camelcase */
          markers: [
            {
              id: "current-location",
              latitude: 25,
              longitude: 25,
              icon: null,
              "visible?": true,
              size: "medium",
              tooltip: null,
              z_index: 1,
              label: null
            }
          ],
          default_center: {
            latitude: 0,
            longitude: 0
          },
          height: 630,
          dynamic_options: {},
          layers: { transit: false },
          auto_init: true,
          bound_padding: null
        }}
        /* eslint-enable typescript/camelcase */
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
