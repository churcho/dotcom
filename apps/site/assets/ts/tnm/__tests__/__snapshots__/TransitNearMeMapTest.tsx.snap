// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders with initial markers 1`] = `<div />`;

exports[`it renders without initial markers 1`] = `<div />`;
